File: common/src/main/kotlin/db/models/Electives.kt
package th.ac.bodin2.electives.db.models

import org.jetbrains.exposed.dao.id.IdTable
import org.jetbrains.exposed.sql.javatime.datetime

/**
 * Electives are a collection of subjects that students can choose from.
 * Each elective can be associated with a team, allowing only students from that team to access the elective to enroll in its subjects.
 */
object Electives : IdTable<Int>("electives") {
    override val id = integer("id").entityId()

    val name = varchar("name", 255)

    /**
     * The team that this elective belongs to. An elective does not need to be associated with a team.
     *
     * If set, only users who belong to this team can access this elective.
     */
    val team = reference("team_id", Teams).nullable()

    val startDate = datetime("start_date").nullable()
    val endDate = datetime("end_date").nullable()

    override val primaryKey = PrimaryKey(id)
}

File: web/src/utils/electives.ts
import type { Elective } from '../api'

export function electiveSorter(a: Elective, b: Elective): number {
    if (a.endDate && b.endDate) {
        const diff = a.endDate.getTime() - b.endDate.getTime()
        if (diff !== 0) return -diff
    }

    if (a.startDate && b.startDate) {
        const diff = a.startDate.getTime() - b.startDate.getTime()
        if (diff !== 0) return diff
    }

    if (!a.startDate && b.startDate) {
        return 1
    }

    if (a.startDate && !b.startDate) {
        return -1
    }

    if (!a.endDate && b.endDate) {
        return 1
    }

    if (a.endDate && !b.endDate) {
        return -1
    }

    if (a.teamId && !b.teamId) {
        return -1
    }

    if (!a.teamId && b.teamId) {
        return 1
    }

    if (a.name < b.name) {
        return -1
    }

    if (a.name > b.name) {
        return 1
    }

    return 0
}
